fluent-bundle/src/resolve.rs

```diff
@@ -13,271 +13,294 @@ use std::hash::{Hash, Hasher};
 
 use super::bundle::FluentBundle;
 use super::entry::GetEntry;
+use super::types::DisplayableNode;
 use super::types::FluentValue;
 use fluent_syntax::ast;
 use fluent_syntax::unicode::unescape_unicode;
 
-#[derive(Debug, PartialEq)]
+#[derive(Debug, PartialEq, Clone)]
 pub enum ResolverError {
-    None,
+    Reference(String),
+    MissingDefault,
+    Argument(String),
     Value,
     Cyclic,
 }
 
 /// State for a single `ResolveValue::to_value` call.
-pub struct Env<'env> {
+pub struct Scope<'scope> {
     /// The current `FluentBundle` instance.
-    pub bundle: &'env FluentBundle<'env>,
+    pub bundle: &'scope FluentBundle<'scope>,
     /// The current arguments passed by the developer.
-    pub args: Option<&'env HashMap<&'env str, FluentValue>>,
+    pub args: Option<&'scope HashMap<&'scope str, FluentValue<'scope>>>,
+    /// Local args
+    pub local_args: Option<HashMap<&'scope str, FluentValue<'scope>>>,
     /// Tracks hashes to prevent infinite recursion.
     pub travelled: RefCell<Vec<u64>>,
+    /// Track errors accumulated during resolving.
+    pub errors: Vec<ResolverError>,
 }
 
-impl<'env> Env<'env> {
+impl<'scope> Scope<'scope> {
     pub fn new(
-        bundle: &'env FluentBundle,
-        args: Option<&'env HashMap<&'env str, FluentValue>>,
+        bundle: &'scope FluentBundle<'scope>,
+        args: Option<&'scope HashMap<&str, FluentValue>>,
     ) -> Self {
-        Env {
+        Scope {
             bundle,
             args,
+            local_args: None,
             travelled: RefCell::new(Vec::new()),
+            errors: vec![],
         }
     }
 
-    fn track<F>(&self, identifier: &str, action: F) -> Result<FluentValue, ResolverError>
+    pub fn track<F>(&mut self, entry: DisplayableNode<'scope>, mut action: F) -> FluentValue<'scope>
     where
-        F: FnMut() -> Result<FluentValue, ResolverError>,
+        F: FnMut(&mut Scope<'scope>) -> FluentValue<'scope>,
     {
         let mut hasher = DefaultHasher::new();
-        identifier.hash(&mut hasher);
+        entry.id.hash(&mut hasher);
```

Comment:
Should we just `(entry.id, entry.attribute).hash()`? That should do more or less the same thing. It _should_ produce the same hash, but I'm not 100% sure. Either way, we just need hash consistency here.